pkg/vere/io/http.c

```diff
@@ -546,6 +553,57 @@ _http_req_dispatch(u3_hreq* req_u, u3_noun req)
   }
 }
 
+
+/* _http_cache_scry_cb(): insert scry result into noun cache
+*/
+static void
+_http_cache_scry_cb(void* vod_p, u3_noun nun)
+{
+  u3_httd* htd_u = vod_p;
+
+  // XX need both the scry path and the noun
+  u3_noun pax = XX;
+
+  u3h_put(htd_u->nax_p, pax, u3k(nun));
+  u3z(nun); // XX should the k and z cancel out?
+}
+
+
+/* _http_req_cache(): attempt to serve http request from cache
+*/
+static c3_o
+_http_req_cache(u3_hreq* req_u)
+{
+  c3_assert(u3_rsat_init == req_u->sat_e);
+  req_u->sat_e = u3_rsat_plan;
+
+  u3_httd* htd_u = req_u->hon_u->htp_u->htd_u;
+
+  u3_noun url = _http_vec_to_atom(req_u->rec_u->path);
+  u3_weak sac = u3h_get(htd_u->sax_p, url);
+  if ( u3_none == sac ) {
+    return c3n;
+  }
+
+  u3_weak nac = u3h_get(htd_u->nax_p, sac);
+  if ( u3_none == nac ) {
+    // noun not in cache; scry it
+    u3_pier_peek_last(htd_u->car_u.pir_u, u3_nul, c3__ex,
+                      u3_nul, sac, htd_u, _http_cache_scry_cb);
+    return c3n; // XX return c3y, and respond async in _http_cache_scry_cb?
```

Comment:
currently this returns c3n, so `_http_rec_accept` will forward the request to Eyre as usual while the scry occurs async (and hopefully, by the time the URL is hit again, the scry will have completed). Would it be better to delay the response until the scry completes?

Relatedly, do we need to worry about multiple concurrent requests to the same URL? Currently this will result in redundant scries, which is wasteful, but idk if it's wasteful enough to justify additional complexity.

Reply:
Vere should not inject requests for URLs that have mappings to scry paths as Arvo events.  It should just scry.  Otherwise, every incoming request will trigger an event, which would put unneeded load on the ship.

Reply:
As for multiple concurrent requests, I would say that's not a huge priority to handle well, since the likelihood of many requests all coming in before the scry result gets cached is fairly low.

An optimization beyond that would be to keep a list of open HTTP requests for each pending scry request, then whenever a scry finishes, Vere sends the result to all the open requests and closes them.  That would make sure a sudden flurry of requests for the same path doesn't end up causing several repeated Arvo scry calls.

I don't think that optimization is worth building for this first version.  Better to stick with the simpler version, which I think will work fine.

Reply:
> Vere should not inject requests for URLs that have mappings to scry paths as Arvo events. It should just scry.

How do I scry without going through Arvo? grepping around, I can see a lot of `*peek` functions, but idk which one to use.

Reply:
I meant it should call something like `u3v_peek()` to call Arvo's `+peek` arm.  This does use the real Arvo core, but it doesn't inject an Arvo *event*, which would call Arvo's `+poke` arm, producing effects and a mutant Arvo state.  Events must also be written to disk, whereas calls to Arvo's `+peek` arm don't change any persistent state, so their inputs don't need to be written to disk.

Reply:
To address your original question more directly:
> _http_rec_accept will forward the request to Eyre as usual while the scry occurs async

I think this should *not* forward requests to Eyre at all if we're also scrying for them.  If we run both the new scry-based pathway and the current event-based pathway, then we haven't reduced load on the publishing ship, because each incoming HTTP request will still trigger a disk write by injecting it as an Arvo event.

I would note that this code is running in the Urth process, not the Mars process that runs Arvo, so both a scry request and an Arvo event are async -- they both get sent over IPC to the Mars process.